tokio/tests/sync_rwlock.rs

```diff
@@ -0,0 +1,163 @@
+#![warn(rust_2018_idioms)]
+
+use std::sync::Arc;
+
+use futures_util::future::FutureExt;
+use futures_util::stream::StreamExt;
+use futures_util::try_future::TryFutureExt;
+use futures_util::{future, stream};
+
+use tokio::sync::{oneshot, RwLock};
+use tokio_test::task::spawn;
+use tokio_test::{assert_pending, assert_ready};
+
+// multiple reads should be Ready
+#[test]
+fn read_shared_ready() {
+    let rwlock = RwLock::new(100);
+    let mut t1 = spawn(rwlock.read());
+    assert_ready!(t1.poll());
+    let mut t2 = spawn(rwlock.read());
+    assert_ready!(t2.poll());
+}
+
+// when there is an active shared ownded, exclusive access should not be possible
+#[test]
+fn write_shared_pending() {
+    let rwlock = RwLock::new(100);
+    let mut t1 = spawn(rwlock.read());
+    let _g1 = assert_ready!(t1.poll());
+    let mut t2 = spawn(rwlock.write());
+    assert_pending!(t2.poll());
+}
+
+// Acquire an RwLock nonexclusively by a single task
+#[tokio::test]
+async fn read_uncontested() {
+    let rwlock = RwLock::new(100);
+    let result = *rwlock.read().await;
+
+    assert_eq!(result, 100);
+}
+
+// Acquire an RwLock nonexclusively by two different tasks simultaneously .
+#[tokio::test]
+async fn read_shared() {
+    let rwlock = RwLock::new(100);
+
+    let (tx0, rx0) = oneshot::channel::<()>();
+    let (tx1, rx1) = oneshot::channel::<()>();
+    let task0 = rwlock.read().then(move |guard| {
+        tx1.send(()).unwrap();
+        rx0.map(move |_| *guard)
+    });
+    let task1 = rwlock.read().then(move |guard| {
+        tx0.send(()).unwrap();
+        rx1.map(move |_| *guard)
+    });
+
+    let result = future::join(task0, task1).await;
+
+    assert_eq!(result, (100, 100));
+}
+
+// Attempt to acquire an RwLock for reading that already has a writer
+#[tokio::test]
+async fn read_contested() {
+    let rwlock = RwLock::<u32>::new(0);
+    let (tx0, rx0) = oneshot::channel::<()>();
+
+    let task1 = rwlock.write().then(move |mut guard| {
+        *guard += 5;
+        rx0.map_err(|_| {
+            drop(guard);
+        })
+    });
+
+    let task2 = rwlock.read().map(|guard| *guard);
+    let task3 = rwlock.read().map(|guard| *guard);
+    let task4 = async { tx0.send(()) };
+
+    let result = future::join4(task1, task2, task3, task4).await;
```

Comment:
I think this test is racy — there's no guarantee that the write lock is held _first_ and that the read locks have to wait for it. In fact, in this test, it might be that the futures all access the critical section sequentially anyway.

Reply:
This is a _great_ candidate for a loom test though.

Reply:
so even if the first task is the first one being polled that doesn't guarantee that it's the first one trying to acquire the write lock? Do you want to change it into a loom test?

Reply:
I don't think `join4` (or any `join`) makes any promises about the order in which it'll poll the underlying futures, which I think this (and the other) tests relies on. As @asomers mentioned above, it's probably to explicitly use polling to do these tests where you rely on specific scheduling.

I think adding a separate file with loom tests (this one being one of them) is a great idea.

Reply:
yeah after reading about it, loom seems designed for these types of tests thanks.
Moved the racy tests to a separate file with loom testing!